codebleu/test_codebleu.py

```diff
@@ -0,0 +1,180 @@
+import json
+import unittest
+import collections
+import codebleu_metric as cbl
+import codebleu.graph_generator.graphgenerator as gg
+import networkx as nx
+from graph_generator.graphgenerator import AstGraphGenerator
+from graph_generator.graphgenutils import prettyprint_graph
+from graph_generator.type_lattice_generator import TypeLatticeGenerator
+from tree_sitter import Language, Parser
+
+
+class TestCodebleuMetric(unittest.TestCase):
+    # ---- DFG part ----
+    parser = Parser()
+    PY_LANGUAGE = Language('codebleu/my-languages.so', 'python')
+    parser.set_language(PY_LANGUAGE)
+    lattice = TypeLatticeGenerator('codebleu/typingRules.json')
+
+    def test_inlist(self):
+        self.assertEqual(cbl._inlist(1, [1, 2, 3]), True)
+        self.assertEqual(cbl._inlist(1, [5, 2, 3]), False)
+        self.assertEqual(cbl._inlist(1, 1), True)
+        self.assertEqual(cbl._inlist(1, 2), False)
+
+    def test_split_identifier(self):
+        self.assertEqual(cbl._split_identifier('1_asd'), [1, 'asd'])
+        self.assertEqual(cbl._split_identifier('1_asd_e'), [1, 'asd_e'])
+        self.assertRaises(Exception, cbl._split_identifier('1_'))
+        self.assertRaises(Exception, cbl._split_identifier('123'))
+        self.assertRaises(Exception, cbl._split_identifier('asd_a'))
+        self.assertRaises(Exception, cbl._split_identifier('___'))
+        self.assertRaises(Exception, cbl._split_identifier('_asd'))
+
+    def test_fix_graph(self):
+        # not sure how to test properly here
+        test_1 = gg.AstGraphGenerator('i=2\ni = i + 1', type_graph=self.lattice).build()
+        test_1_dic = cbl._fix_graph(test_1)
+        self.assertEqual(test_1_dic['edges']['NEXT_USE'], {'3_i': ['14_i'], '14_i': ['10_i']})
+        self.assertEqual(test_1_dic['edges']['OCCURRENCE_OF'], {'3_i': ['4_i'], '10_i': ['4_i'], '14_i': ['4_i']})
+
+    @staticmethod
+    def _split_merge_identifier(myvar):
+        split_myvar = cbl._split_identifier(myvar)
+        update_myvar = str(split_myvar[0] - 1) + '_' + split_myvar[1]
+        return update_myvar
+
+    def test_parse_occurrences(self, g=None):
+        if g is None:
+            snp = gg.AstGraphGenerator('i=2\ni = i + 1', type_graph=self.lattice).build()
```

Comment:
There are several graphs that appear from test to test (this one, the directed graphs from `networkx`). You can move them to function-level fixtures and keep only the actual testing logic in test functions.

Reply:
Function-level fixtures will be run from scratch for each test, so you do not have to care about modifying the graphs in this case.